### PR TITLE
feat(solr-transforms): Add cursor pagination (cursorMark → search_after)

### PR DESCRIPTION
**Pull request by niravpi**
### Description

Adds cursor-based pagination to the Solr→OpenSearch shim, translating cursorMark to search_after. Includes full dual-mode support where both Solr and OpenSearch paginate using their native mechanisms via a combined cursor token.

Changes:

- **cursor-pagination.ts** (new)
  - Request: decodes cursorMark to search_after, parses sort with _id tiebreaker, handles + as space.
  - Response: base64-encodes last hit's sort values as nextCursorMark, returns same token on empty results (end signal). Uses pure JS base64 (GraalVM has no btoa/Buffer/Java.type).
  - Works in both single-mode and dual-mode — dual-mode token splitting/merging is handled by the Java shim layer.

- **context.ts** — Add targetName and dualMode fields to RequestContext/ResponseContext, parsed from _targetName/_dualMode injected by the shim proxy.

- **MultiTargetRoutingHandler.java** — Dual-mode cursor pagination support:
  - Request side: decodes combined cursor token and rewrites cursorMark per target — Solr gets its native token, OpenSearch gets the encoded sort values.
  - Response side: merges nextCursorMark from both targets into a combined base64 token {solr: "<native>", os: "<sort_values>"}.
  - Refactored mergeCursorTokens to reduce cognitive complexity (extracted isPaginationEnded and logPartialCursorWarning helpers) — fixes SonarQube S3776.
  - Deep copies request map when cursorMark is present in dual mode to avoid cross-target URI contamination.
  - Inject _targetName and _dualMode into request map so transforms can detect dual-target mode.
  - Fix: pass original requestMap (not modified copy) as originalRequestMap to dispatchToTarget — the modified copy was breaking response transform bundling.

- **MultiTargetRoutingHandlerCursorTest.java** (new) — 21 Java unit tests covering extractQueryParam, replaceQueryParam, rewriteCursorMarkForTarget, and mergeCursorTokens (token combining, end-of-pagination detection, invalid/partial tokens, skip when one missing, response body preservation, edge cases).

- **ShimProxyTest.java** — Added dualTarget_cursorMarkMerge integration test covering the full Netty pipeline for cursor token merging in dual-target mode.

- **hits-to-docs.ts** — Add id from hit._id metadata (was missing from source), force start=0 for cursor responses, improve field wrapping heuristic.

- **registry.ts** — Register cursor-pagination in request (after query-q) and response (before hits-to-docs) pipelines.

- **Test framework** — Add requestSequence with {{nextCursorMark}} substitution for multi-step e2e tests (test-types.ts, TestCaseDefinition.java, TransformationShimE2ETest.java).

- **run.sh** — Minor enhancement for shim build support.

- **Integration test logging** — Updated e2e test runner to show detailed logs for easier debugging.

### Architecture Note

The cursor token merge/split logic lives in the Java shim layer (MultiTargetRoutingHandler) because it requires cross-target visibility — access to both Solr's and OpenSearch's responses simultaneously. The TypeScript transformation layer only sees one target's response at a time, so it handles the Solr↔OpenSearch query translation while the shim handles multi-
target orchestration.

### Known Limitations

- **Traffic Replay:** Replaying captured Solr traffic with cursorMark through the traffic replayer is not supported — Solr's native cursor tokens are opaque binary and cannot be decoded to search_after values. This is a separate concern from the live shim proxy.
- **Custom uniqueKey:** The shim assumes id is the Solr uniqueKey (Solr's default). Custom uniqueKey field names are not auto-detected. See limitation.md for details.

### Testing

- **Unit tests (TypeScript):** 33 cursor-specific tests (b64 roundtrip, sort parsing, cursor encode/decode, request/response transforms, dual-mode match) — 418 total passing
- **Unit tests (Java):** 21 new cursor tests + 1 integration test (extractQueryParam, replaceQueryParam, rewriteCursorMarkForTarget, mergeCursorTokens with end detection, invalid tokens, edge cases, full Netty pipeline merge) — all passing

### CODE REVIEW

File: TrafficCapture/transformationShim/src/main/java/org/opensearch/migrations/transform/shim/netty/MultiTargetRoutingHandler.java [hunk 1]
```diff
@@ -433,6 +451,122 @@ private FullHttpResponse buildPrimaryResponse(TargetResponse primary) {
...
+                        osToken = nextCursor.toString();
+                    } else {
+                        solrToken = nextCursor.toString();
+                    }
+                }
+            }
+        }
+
+        // Only merge if both targets returned cursor tokens
+        if (solrToken == null || osToken == null) return response;
```
[the remark this answers is not in the document]
Comment (niravpi) [reply TrafficCapture/transformationShim/src/main/java/org/opensearch/migrations/transform/shim/netty/MultiTargetRoutingHandler.java]: Added a WARN log when only one target returns a cursor token: "Only one target returned a cursor token (solr=present/missing, os=present/missing), skipping merge". The skip itself is correct (can't merge partial tokens), but now it's visible for debugging.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/cursor-pagination.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+const CURSOR_MARK_START = '*';
+
+/** Solr uniqueKey field name — maps to OpenSearch's _id. */
+const SOLR_UNIQUE_KEY = 'id';
+const OS_UNIQUE_KEY = '_id';
+
+const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+
+/** UTF-8 encode a string into a byte array (GraalVM-safe, no TextEncoder). */
+function toUTF8Bytes(str: string): number[] {
```
Comment (shuklas) [inline TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/cursor-pagination.ts]: Why using TextEncoder wouldn't have been safe in GraalVM?

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/cursor-pagination.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  SOLR_UNIQUE_KEY, OS_UNIQUE_KEY };
+
+export const request: MicroTransform<RequestContext> = {
+  name: 'cursor-pagination',
+  match: (ctx) => ctx.params.has('cursorMark'),
+  apply: (ctx) => {
+    const cursorMark = ctx.params.get('cursorMark')!;
+
+    // cursorMark and start are mutually exclusive — remove from if present
+    ctx.body.delete('from');
```
[the remark this answers is not in the document]
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/cursor-pagination.ts]: Actually from is correct here. Because at this point in the pipeline, the earlier start-rows transform has already translated Solr's start param into OpenSearch's from in the request body. So we're deleting the OpenSearch body field, not the Solr param.

File: TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/cursor-pagination.ts (hunk 2 above)
Comment (niravpi) [reply TrafficCapture/SolrTransformations/transforms/src/solr-to-opensearch/features/cursor-pagination.ts]: TextEncoder/TextDecoder are Web APIs (part of the WHATWG Encoding spec), not part of the ECMAScript standard. GraalVM's JS engine only provides ECMAScript globals in its scope such as Array, ArrayBuffer, Map, Set, Promise, etc. TextEncoder is not among them (same reason btoa, atob, and Buffer aren't available). Since we run inside a polyglot Context and don't control the engine configuration, we use a pure ES2020 manual UTF-8 encoder that works in any standard JS engine.

File: TrafficCapture/transformationShim/src/main/java/org/opensearch/migrations/transform/shim/netty/MultiTargetRoutingHandler.java [hunk 4]
```diff
@@ -462,6 +486,140 @@ private FullHttpResponse buildPrimaryResponse(TargetResponse primary) {
...
+            String combinedToken = buildCombinedToken(solrToken, osToken);
+            return rewriteResponseWithToken(response, combinedToken);
+        } catch (Exception e) {
+            log.warn("Failed to merge cursor tokens: {}", e.getMessage());
+            return response;
+        }
+    }
+
+    private String extractCursorToken(Map<String, TargetResponse> allResponses, String targetKey) {
+        for (var entry : allResponses.entrySet()) {
```
[the remark this answers is not in the document]
Comment (niravpi) [reply TrafficCapture/transformationShim/src/main/java/org/opensearch/migrations/transform/shim/netty/MultiTargetRoutingHandler.java]: ack, updated in latest revision
